core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -337,7 +338,7 @@ object Pull extends PullLowPriority {
     }
   }
 
-  private abstract class Bind[F[_], O, X, R](val step: Pull[F, O, X]) extends Pull[F, O, R] {
+  private abstract class Bind[+F[_], +O, X, +R](val step: Pull[F, O, X]) extends Pull[F, O, R] {
```

Comment:
With no changes to this encoding, Dotty complains about 8 uncheckable runtime type checks due to variance mismatches. I was able to remove all of those by changing variance but more casts were needed. I minimized the number of such casts but there might be room here for more improvements or perhaps lurking bugs in Dotty.